src/main/java/org/opensearch/knn/index/vectorvalues/KNNMergeVectorValues.java

```diff
@@ -0,0 +1,120 @@
+            throw new UnsupportedOperationException("Cannot merge vectors encoded as [" + fieldInfo.getVectorEncoding() + "] as FLOAT32");
+        }
+        final List<KNNVectorValuesSub<FloatVectorValues>> subs = new ArrayList<>();
+        for (int i = 0; i < mergeState.knnVectorsReaders.length; i++) {
+            KnnVectorsReader knnVectorsReader = mergeState.knnVectorsReaders[i];
+            if (knnVectorsReader != null) {
+                FloatVectorValues values = knnVectorsReader.getFloatVectorValues(fieldInfo.getName());
+                if (values != null) {
+                    final Bits liveDocs = mergeState.liveDocs[i];
+                    final int liveDocsCardinality = cardinality(liveDocs, Math.toIntExact(values.cost()));
+                    subs.add(new KNNVectorValuesSub<>(mergeState.docMaps[i], values, liveDocsCardinality));
+                }
+            }
+        }
+        return new KNNVectorValuesIterator.MergeFloat32VectorValuesIterator(subs, mergeState);
+    }
+
+    /**
+     * Gets list of {@link KNNVectorValuesSub} for {@link ByteVectorValues} from a merge state. This can be further
+     * used to create an iterator for getting the docs and its vector values
+     * @param fieldInfo
+     * @param mergeState
+     * @return List of KNNVectorSub
+     * @throws IOException
+     */
+    public static KNNVectorValuesIterator.MergeByteVectorValuesIterator mergeByteVectorValues(FieldInfo fieldInfo, MergeState mergeState)
+        throws IOException {
+        assert fieldInfo != null && fieldInfo.hasVectorValues();
+        if (fieldInfo.getVectorEncoding() != VectorEncoding.BYTE) {
+            throw new UnsupportedOperationException("Cannot merge vectors encoded as [" + fieldInfo.getVectorEncoding() + "] as BYTE");
+        }
+        final List<KNNVectorValuesSub<ByteVectorValues>> subs = new ArrayList<>();
+        for (int i = 0; i < mergeState.knnVectorsReaders.length; i++) {
+            KnnVectorsReader knnVectorsReader = mergeState.knnVectorsReaders[i];
+            if (knnVectorsReader != null) {
+                ByteVectorValues values = knnVectorsReader.getByteVectorValues(fieldInfo.getName());
+                if (values != null) {
+                    final Bits liveDocs = mergeState.liveDocs[i];
+                    final int liveDocsCardinality = cardinality(liveDocs, Math.toIntExact(values.cost()));
+                    subs.add(new KNNVectorValuesSub<>(mergeState.docMaps[i], values, liveDocsCardinality));
+                }
+            }
+        }
+        return new KNNVectorValuesIterator.MergeByteVectorValuesIterator(subs, mergeState);
+    }
+
+    private static int cardinality(final Bits liveDocs, final int defaultCount) {
+        if (liveDocs == null) {
+            return defaultCount;
+        }
+
+        if (liveDocs instanceof FixedBitSet) {
+            return ((FixedBitSet) liveDocs).cardinality();
+        }
```

Comment:
why this special case for fixedbitset and not for `SparseFixedBitSet` ?

Reply:
I don't think its being used. I was looking at the usages of [LiveDocsFormats write](https://github.com/apache/lucene/blob/53d1c2bd2fb3e6b9da590bee360996dbbdc8ea34/lucene/core/src/java/org/apache/lucene/codecs/LiveDocsFormat.java#L44) 
- [DocumentWritersPerThread](https://github.com/apache/lucene/blob/53d1c2bd2fb3e6b9da590bee360996dbbdc8ea34/lucene/core/src/java/org/apache/lucene/index/DocumentsWriterPerThread.java#L661)
- [Pending Deletes](https://github.com/apache/lucene/blob/53d1c2bd2fb3e6b9da590bee360996dbbdc8ea34/lucene/core/src/java/org/apache/lucene/index/PendingDeletes.java#L194)

read only returns FixedBits. Moreover while debugging I did not see SparseFixedBitSet usage. I can change it to `instanceof BitSet` to be safe

Reply:
I am not super confident with this and also the next loop to find the live docs. I would check with lucene community first before making any assumptions here to avoid any possible regression, since this is on the write path.

Reply:
> community first before making any assumptions

Can you elaborate about the concerns with BitSet usage, as in performance or functionality? cardinality is supposed to return set bits and lucene seems to have optimized it for Fixed and Sparse with bit shifting

> the next loop to find the live docs

For this is the concern performance or functionality?

Asking because we should do our due diligence before reaching out to the lucene community

Reply:
@shatejas I think we don't need to reach out to Lucene, I thought of the case where this bitset count will fail and you can try this out with your cluster.

Live docs tells the total number of docs which are not deleted on the segment. Now a doc may or may not contain the vector field. So consider a case where there are 3 docs in the segment.
doc1 : has vector field : doc is deleted
doc2: has no vector field : doc is not deleted
doc3: has vector field : doc is not deleted.

With the above case the bitset will give live docs as 2 but for the vector field the live docs is only 1. 

Hence, using the bitset.cardinality() will never give you right number of live docs.

and this is the reason why lucene doesn't use this during segment merges during the training.

cc: @jmazanec15 and @heemin32